### ISSUE TITLE
Support multi-dex

### ISSUE THREAD

**Issue by radare** [radare/radare2#5133]

**Comment by radare** [comment]
first we need to load imports and fields properly before going full-multi-dex way

**Comment by h4ng3r** [comment]
we can create a  apk:// or  multidex:// similar to the Java  zipall://.

**Comment by radare** [comment]
👍

**Comment by Maijin** [comment]
r2 -L, apk:// already exist, just need to be adapted.
* rw_  zip      Open zip files [apk|ipa|zip|zipall]://[file//path] (BSD)

**Comment by h4ng3r** [comment]
Perfect! I'll do it after fix the dex parser

**Comment by Maijin** [comment]
what do you mean, dex parser? Dex is already handled by r2

**Comment by h4ng3r** [comment]
I'm fixing some bugs :)

**Comment by Maijin** [comment]
ah ok like this https://github.com/radare/radare2/issues/5317 :3

**Comment by h4ng3r** [comment]
For the moment I fixed:
* Parsing bin->strings for UTF-8 classes, methods, fields.
* Add signature to all methods and fields, so there is no need to use an index
* Minor parsing errors that I don't remember

**Comment by h4ng3r** [comment]
I'm working on this issue.

Is quite easy to adapt "r_io_zip_open" to append all classes*.dex to the same RIOZipFileObj object. But I'm not sure which is the best way to handle it in the bin_dex.c. There all main functions get a "RBinFile" object, one idea is change RBinDexObj* holded on arch->o->bin_obj for a RList and then change all functions to do the same but for all dex files.
Another idea is to get a plugin between zip and dex parser, this plugin could create a RBInObject for each dex files, pass to the dex parser and then merge all info. Seems a more elegant solution to me but I'm not sure.
What do you think?

**Comment by radare** [comment]
We should probably load them all into a single binobj. Not sure how its done for java but would be good to use the same method that it uses to load all the classes in a jar at once. Also im not sure if there are tests for this

**Comment by h4ng3r** [comment]
Not sure if it's working for Java:
```
$ r2 apktool_2.2.0.jar
 -- Find hexpairs with '/x a0 cc 33'
[0x00000000]> ic

[0x00000000]> is
[Symbols]

0 symbols

[0x00000000]> q
$ r2 jar://apktool_2.2.0.jar
Do you want to print 1782 lines? (y/N)
Cannot open 'jar://apktool_2.2.0.jar'
$
```

For others like IPA, it is only loading one file from the "zip". 

I've a working PoC that reads 2 dexs and puts both in the same RBin. The bin_dex compares the size of the first DexObj and detects that are two files. I think that with a little refactoring it can be done easily.

**Comment by h4ng3r** [comment]
In fact, parsing is easy because you always know which Dex are you parsing. The problem is when disas or anal need some info, now they are calling getoffset or getname and passing class_id, type_id, string_id or method_id. With multi-dex the method_id is not unique, so we need to also pass the DEX file or maybe the address so bin_dex can get which file is it.

**Comment by Maijin** [comment]
for jar it's `r2 zip://`
for apk it's `apk://`
there is also `zipall://`, `ipa://` etc. check `r2 -L`

**Comment by Maijin** [comment]
I've already said that previously btw

**Comment by Maijin** [comment]
![image](https://cloud.githubusercontent.com/assets/1408600/18818585/5574c078-837d-11e6-8eb0-be93e01f2388.png)

**Comment by h4ng3r** [comment]
@Maijin I know that, the problems is how because there are multi solutions and I'm trying to decide with @pancake which is the best one.

**Comment by radare** [comment]
not gonna happen either for 1.5 . @h4ng3r ¬¬U

**Comment by h4ng3r** [comment]
@radare dont think so... Multidex should not be implemented alone it should be part of some multibin for any arch. I checked zip and not sure if java is doing it or how is doing it. For me dex arena is more important.

**Comment by ArbitraryRW** [comment]
@h4ng3r @radare Did you ever agree on the best solution? I would be happy to help out.

**Comment by radare** [comment]
i have been working on the rbin refactoring on the previous release, and i  have plans to continue improving forward in this direction. there are several more things in DEX to be supported and improved in the parser. do you have telegram? maybe we can coordinate some of those efforts

**Comment by ArbitraryRW** [comment]
Absolutely, more than happy to coordinate. I unfortunately don't have telegram but I have signal / slack

**Comment by radare** [comment]
slack may work, but there's no slack channel for r2. i dont know if it is possible to bridge irc/telegram/matrix with slack in a simple way

**Comment by radare** [comment]
cc @wargio

**Comment by Surendrajat** [comment]
+1 for this. Quite an important feature as most of the big apks have more than one dex files anyways.

**Comment by ret2libc** [comment]
This issue has been moved from radareorg/radare2 to radareorg/ideas as we are trying to clean our backlog and this issue has probably been created a long while ago. This is an effort to help contributors understand what are the actionable items they can work on, prioritize issues better and help users find active/duplicated issues more easily. If this is not an enhancement/improvement/general idea but a bug, feel free to ask for re-transfer to main repo. Thanks for your understanding and contribution with this issue.
